datafusion-examples/examples/table_sample.rs

```diff
@@ -0,0 +1,1353 @@
+            Arc::clone(&children[0]),
+            self.lower_bound,
+            self.upper_bound,
+            self.with_replacement,
+            self.seed,
+        )?))
+    }
+
+    fn execute(
+        &self,
+        partition: usize,
+        context: Arc<TaskContext>,
+    ) -> Result<SendableRecordBatchStream> {
+        let input_stream = self.input.execute(partition, context)?;
+        let baseline_metrics = BaselineMetrics::new(&self.metrics, partition);
+
+        Ok(Box::pin(SampleExecStream {
+            input: input_stream,
+            sampler: self.create_sampler(partition)?,
+            baseline_metrics,
+        }))
+    }
+
+    fn metrics(&self) -> Option<MetricsSet> {
+        Some(self.metrics.clone_inner())
+    }
+
+    fn partition_statistics(&self, partition: Option<usize>) -> Result<Statistics> {
+        let input_stats = self.input.partition_statistics(partition)?;
+
+        // Apply sampling ratio to statistics
+        let mut stats = input_stats;
+        let ratio = self.upper_bound - self.lower_bound;
+
+        stats.num_rows = stats
+            .num_rows
+            .map(|nr| (nr as f64 * ratio) as usize)
+            .to_inexact();
+        stats.total_byte_size = stats
+            .total_byte_size
+            .map(|tb| (tb as f64 * ratio) as usize)
+            .to_inexact();
+
+        Ok(stats)
+    }
+}
+
+/// Stream for the SampleExec operator
+struct SampleExecStream {
+    /// The input stream
+    input: SendableRecordBatchStream,
+    /// The sampling method
+    sampler: Box<dyn Sampler>,
+    /// Runtime metrics recording
+    baseline_metrics: BaselineMetrics,
+}
+
+impl Stream for SampleExecStream {
+    type Item = Result<RecordBatch>;
+
+    fn poll_next(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Self::Item>> {
+        let poll = self.input.poll_next_unpin(cx);
+        let baseline_metrics = &mut self.baseline_metrics;
+
+        match poll {
+            Poll::Ready(Some(Ok(batch))) => {
+                let start = baseline_metrics.elapsed_compute().clone();
+                let result = self.sampler.sample(&batch);
+                let _timer = start.timer();
+                Poll::Ready(Some(result))
+            }
+            Poll::Ready(Some(Err(e))) => Poll::Ready(Some(Err(e))),
+            Poll::Ready(None) => Poll::Ready(None),
+            Poll::Pending => Poll::Pending,
+        }
+    }
+}
+
+impl RecordBatchStream for SampleExecStream {
+    fn schema(&self) -> SchemaRef {
+        self.input.schema()
+    }
+}
+
+/// Query planner to produce a LogicalPlan from AST with table sampling
+struct TableSamplePlanner<'a, S: ContextProvider> {
+    context_provider: &'a S,
+}
+
+fn evaluate_number<
+    T: FromStr + Add<Output = T> + Sub<Output = T> + Mul<Output = T> + Div<Output = T>,
+>(
+    expr: &ast::Expr,
+) -> Option<T> {
+    match expr {
+        ast::Expr::BinaryOp { left, op, right } => {
```

Comment:
most of the rest of DataFusion would use an `Expr` here and reuse the existing simplification implementation (e.g. https://github.com/apache/datafusion/blob/a8d5016b97481d837857036ddd450574455a1b4e/datafusion-examples/examples/expr_api.rs#L162)

Reply:
Tried to rework this functionality using a more generic expression simplified. Since this change is quite large, I would prefer to submit it as a different follow-up PR.